lib/Sema/CSDiag.cpp

```diff
@@ -5410,13 +5415,79 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
+    auto candidate = calleeInfo[0];
+    // We have an initializer expression for some type (might be generic),
+    // before giving up trying to diagnose this situation, let's double
+    // check if we have any unresolved archetypes, and try to diagnose that.
+    // This might happen in the situation like following:
+    // struct S<A, B> { init(a: A) {} } and call S(a: 1),
+    // initializer is going to be (A) -> S<A, B>, A is inferred from
+    // initializer argument and B going to be unspecified.
+    auto decl = candidate.getDecl();
+    // If the candidate is a valid constructor declaration, let's proceed.
+    if (decl && !decl->isInvalid() && isa<ConstructorDecl>(decl)) {
+      auto resultType = candidate.getResultType();
+      // We can only try to diagnose bound generic types.
+      if (auto BGT = resultType->getAs<BoundGenericType>()) {
+        llvm::SmallSet<ArchetypeType *, 2> missingParams;
+        for (auto &genericParam : BGT->getGenericArgs()) {
+          if (auto archetype = genericParam->getAs<ArchetypeType>()) {
+            // First, let's consider all of the generic arguments missing.
+            missingParams.insert(archetype);
+          }
+        }
+
+        std::function<void(Type)> recordArchtype = [&](Type argumentType) {
+          // Syntax Sugar type e.g. Optional or Array.
+          if (auto SST = dyn_cast<SyntaxSugarType>(argumentType.getPointer()))
+            argumentType = SST->getBaseType();
+
+          // Handles associative collections e.g. Dictionary or Set.
+          if (auto BGT = argumentType->getAs<BoundGenericType>()) {
+            for (auto &generic : BGT->getGenericArgs())
+              recordArchtype(generic);
+            return;
+          }
+
+          if (auto archetype = argumentType->getAs<ArchetypeType>()) {
+            // We've found one of the archetypes on the generic parameters,
+            // let's remove it from missing because it can be inferred
+            // from argument type in initializer.
+            missingParams.erase(archetype);
+          }
+        };
+
+        auto argumentType = candidate.getArgumentType().getPointer();
+        // A single unlabeled argument.
+        if (auto parenType = dyn_cast<ParenType>(argumentType))
+          recordArchtype(parenType->getUnderlyingType());
+
+        // A single labeled argument or multiple arguments.
+        if (auto tupleType = dyn_cast<TupleType>(argumentType)) {
+          for (auto &arg : tupleType->getElements())
+            recordArchtype(arg.getType());
+        }
+
+        if (missingParams.size() > 0) {
+          diagnoseUnboundArchetype(*missingParams.begin(), fnExpr);
+          return true;
+        }
+      }
+    }
+  }
+
   // If we found an exact match, this must be a problem with a conversion from
   // the result of the call to the expected type.  Diagnose this as a
   // conversion failure.
   if (calleeInfo.closeness == CC_ExactMatch)
     return false;
-  
+
```

Comment:
Can you try to avoid whitespace cleanup like this?

It makes git blame history more challenging to follow.